src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +151,150 @@ internal class DiagnosticComputer
+                // Step 4:
+                //  - Execute the core 'computeTask' for diagnostic computation.
+                return await computeTask.ConfigureAwait(false);
+            }
+            finally
+            {
+                // Step 5:
+                //  - Remove the 'computeTask' from the list of current executing high priority tasks.
+                lock (s_gate)
+                {
+                    Debug.Assert(s_highPriorityComputeTasks.Contains(computeTask));
+                    s_highPriorityComputeTasks = s_highPriorityComputeTasks.Remove(computeTask);
+                }
+            }
+
+            static void CancelNormalPriorityTasks()
+            {
+                // Fire cancellation on the current token source and replace it with a new token source for future normal priority requests.
+                CancellationTokenSource currentCts;
+                lock (s_gate)
+                {
+                    currentCts = s_cancellationSourceForNormalPriorityComputeTasks;
+                    s_cancellationSourceForNormalPriorityComputeTasks = new();
+                }
+
+                currentCts.Cancel();
+                currentCts.Dispose();
+            }
+        }
+
+        private async Task<SerializableDiagnosticAnalysisResults> GetNormalPriorityDiagnosticsAsync(
+            IEnumerable<string> analyzerIds,
+            bool reportSuppressedDiagnostics,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Step 1:
+                //  - Normal priority task must wait for all the executing high priority tasks to complete
+                //    before beginning execution.
+                await WaitForHighPriorityTasksAsync(cancellationToken).ConfigureAwait(false);
+
+                // Step 2:
+                //  - Create the core 'computeTask' for computing diagnostics.
+                var computeTask = CreateComputeTask(cancellationToken);
+
+                try
+                {
+                    // Step 3:
+                    //  - Execute the core 'computeTask' for diagnostic computation.
+                    return await computeTask.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Step 4:
+                    // Attempt to re-execute this cancelled normal priority task by running the loop again,
+                    // unless cancellation was fired on the cancellationToken passed into us.
+                    cancellationToken.ThrowIfCancellationRequested();
+                    continue;
+                }
+            }
+
+            Task<SerializableDiagnosticAnalysisResults> CreateComputeTask(CancellationToken cancellationToken)
+            {
+                return Task.Run(async () =>
```

Comment:
>return Task.Run(async () => [](http://example.com/codeflow?start=16&length=27)

Just curious, what's the value in using Task.Run here instead of just returning the task from GetDiagnosticsAsync?